### PR TITLE
Speedloader: Mosin Edition

### PR DESCRIPTION
**Pull request by PixelTheKermit**
## About the PR 
The mosin can now be reloaded with a speedloader
This took too long for what it was

**Screenshots**

https://user-images.githubusercontent.com/85175107/202796953-a4495276-f80e-4e24-a280-08e76ade897d.mp4

**Changelog**

:cl: PixelTK
- fix: You can now use speedloaders on the mosin
- tweak: Speedloaders have replaced the magazines in the syndicate uplink

### CODE REVIEW

**Comment by PixelTheKermit** [comment]
Still need to do stuff to this don't merge pls

**Comment by PixelTheKermit** [comment]
Ok it's ready now

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 1]
```diff
@@ -41,15 +41,43 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
+                component.Entities.Add(bullet);
+                component.Container.Insert(bullet);
+                UpdateBallisticAppearance(ammoComp); // This one is for the speedloader
+            }
+        }
+        else
+        {
+            component.Entities.Add(args.Used);
+            component.Container.Insert(args.Used);
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:74]: This code should be on speedloaders, not here, this is just going to spawn a billion entities.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by JustinTrotter** [comment]
Getting a whole box for 2 TC is a bit much. I think getting a 10 round rifle speed loader would do.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs (hunk 1 above)
Comment (PixelTheKermit) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:74]: tbh I should probably not spawn new bullets and instead just increment "unspawnedcount"

Comment (metalgearsloth) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:74]: Also my guy TryComp exists.

**Comment by PixelTheKermit** [comment]
> Getting a whole box for 2 TC is a bit much. I think getting a 10 round rifle speed loader would do.

I mean, the gun kinda sucks (having to manually cycle and having a slow ass fire rate), and making the ammo cost more would make it not worth buying (it probably wasn't worth buying before at all).

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs (hunk 1 above)
Comment (PixelTheKermit) [reply Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs]: Changed the code to not spawn in a billion entities
also yes, TryComp does exists, good to know

**Comment by Mervill** [comment]
With Magazines you actually get 180 rounds total (6 mags of 30 rounds). Whereas with this PR you would get only 30 rounds (6 clips of 5) which is already a significant reduction.

**Comment by JustinTrotter** [comment]
> With Magazines you actually get 180 rounds total (6 mags of 30 rounds). Whereas with this PR you would get only 30 rounds (6 clips of 5) which is already a significant reduction.

But you only get a single magazine for 2 TC, not a whole box of them.

**Comment by PixelTheKermit** [comment]
> > With Magazines you actually get 180 rounds total (6 mags of 30 rounds). Whereas with this PR you would get only 30 rounds (6 clips of 5) which is already a significant reduction.
> 
> But you currently only get a single magazine for 2 TC, not a whole box of them.

Wrong!
image

**Comment by JustinTrotter** [comment]
> > > With Magazines you actually get 180 rounds total (6 mags of 30 rounds). Whereas with this PR you would get only 30 rounds (6 clips of 5) which is already a significant reduction.
> > 
> > 
> > But you currently only get a single magazine for 2 TC, not a whole box of them.
> 
> Wrong! image

Let me clarify. Currently from the uplink you can buy single magazines and....

Huh. I retract everything.

**Comment by mirrorcult** [comment]
i dont know anything about guns but it doesnt make any sense that these would be easy to reload

**Comment by EmoGarbage404** [comment]
Isnt it a bolt-action rifle? The point is that you crank dat sucka. 😎

**Comment by SpliTistiC** [comment]
> i dont know anything about guns but it doesnt make any sense that these would be easy to reload

'Stripper Clips' have existed since the 1900s for this reason, to make loading rifles easier.

**Comment by mirrorcult** [comment]
we will not be adding Strippers to ss14 that is erp.

**Comment by PixelTheKermit** [comment]
> i dont know anything about guns but it doesnt make any sense that these would be easy to reload

Does anything make sense in space station?

> Isnt it a bolt-action rifle? The point is that you crank dat sucka. 😎

If we're taking about realism, I'm certain the mosin is only supposed to hold 5 bullets, dum space game never playing again :godo:

**Comment by mirrorcult** [comment]
> > i dont know anything about guns but it doesnt make any sense that these would be easy to reload
> 
> Does anything make sense in space station?
> 

these have a design reason to be shitty and hard to reload

**Comment by PixelTheKermit** [comment]
> > > i dont know anything about guns but it doesnt make any sense that these would be easy to reload
> > 
> > 
> > Does anything make sense in space station?
> 
> these have a design reason to be shitty and hard to reload

They're already bad enough, I don't see how giving them speed loading capabilities is going to make them any better. You already have to cycle the gun every shot and wait a second before firing again.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 2]
```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
-        component.Entities.Add(args.Used);
-        component.Container.Insert(args.Used);
+        if (TryComp<BallisticAmmoProviderComponent>(args.Used, out var ammoComp))
+        {
+            if (GetBallisticShots(ammoComp) == 0) return;
+
+            while (GetBallisticShots(component) < component.Capacity && GetBallisticShots(ammoComp) > 0)
+            {
+                var xform = EntityManager.GetComponent<TransformComponent>(uid);
+
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:51]: Don't resolve the same component in loops please, it's very slow.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 3]
```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
 
         if (GetBallisticShots(component) >= component.Capacity) return;
 
-        component.Entities.Add(args.Used);
-        component.Container.Insert(args.Used);
+        if (TryComp<BallisticAmmoProviderComponent>(args.Used, out var ammoComp))
+        {
+            if (GetBallisticShots(ammoComp) == 0) return;
+
+            while (GetBallisticShots(component) < component.Capacity && GetBallisticShots(ammoComp) > 0)
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:48]: Isn't this going to continue to iterate a container over and over? Seems much better to track it yourself.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 4]
```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
+            {
+                var xform = EntityManager.GetComponent<TransformComponent>(uid);
+
+                if (ammoComp.Entities.Count == 0) // If the entity doesn't have any spawned bullets
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:52]: Comment is unnecessary.

File: Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs [hunk 5]
```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
...
+                {
+                    if (GetBallisticShots(component) == 0) // If the gun doesn't have any ammo
```
Comment (metalgearsloth) [inline Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs:54]: Same here, and also refer to my above comment on "unnecessarily iterating the container frequently".

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This is still going to drop ticks, this is why dumping bullets doesn't dump the entire thing.

**Comment by metalgearsloth** [comment]
Just make it do 1 at a time please.

**Comment by PixelTheKermit** [comment]
> Just make it do 1 at a time please.

So make the speedloader just load one bullet every use? Sure.

**Comment by PixelTheKermit** [comment]
I have unlooped the code

**Comment by PixelTheKermit** [comment]
I'll close this, since I don't think it's good enough to be merged

**Pull request closed without merging**